### ISSUE TITLE
Add the "lean bundle" use-case no-restricted-imports

### ISSUE THREAD

**Issue by dozoisch** [eslint/eslint#7532]
From discussions with @ljharb in https://github.com/benmosher/eslint-plugin-import/issues/642 I realised that what I wanted to add in eslint-plugin-import is actually already supported by no-restricted-imports and no-restricted-modules. It just wasn't a very obvious use case in the documentation. 

Let me explain a bit what the use case is! Basically we have a plugin https://github.com/eslint-plugins/eslint-plugin-lean-imports that warns people when they pull in a full module instead of pulling just a single file. Some packages advise people to pull single files in order to have smaller bundles (eg: lodash, react-bootstrap). 

```js
// with "lean-imports/import": [1, [ "react-bootstrap "] ]
import RB from 'react-bootstrap'; // warns
const RB = require('react-bootstrap'); // warns

const Button = require('react-bootstrap/lib/button'); // passes
```

As I said, I can have a similar behavior by setting

```js
{
  "no-restricted-imports": [2, "react-bootstrap"],
  "no-restricted-modules":[2, "react-bootstrap"],
}
```

The issue is just that the error message is not really tailored at that behavior and is really obscure if that is what the wanted behavior was: `'react-bootstrap' import is restricted from being used`. It would be nice if there was a way to specify something more tailored for these packages (like have different "sections" or something along that), or have a message that supports both! 

The message we have in lean-import is the following `Importing the entire lodash library is not permitted, please import the specific functions you need` which tells the user why it errored. 

Also, Adding documentation about that specific use case in the doc of both of these rules could help people learn about it!

I'm willing to do the PRs with these, I'm just not sure of the approach here.

**Comment by ljharb** [comment]
I'd suggest making the schema allow an array of either a string (current), or `{ "name": "<current string>", "message": "<optional message>" }` - that lets it be a smooth backwards-compatible upgrade on top of the current schema, and mirrors `no-restricted-properties` for "message".

**Comment by kaicataldo** [comment]
Woof, that's going to be quite a complicated schema once this lands: https://github.com/eslint/eslint/pull/7433

I'm for this change though. It's consistent with `no-restricted-properties`, and allowing for custom messaging to provide context for these kinds of restrictions is very useful.

**Comment by dozoisch** [comment]
Perfect! I'll start working on it when I get some time today.

**Comment by eslintbot** [comment]
Hi @dozoisch, thanks for the issue. It looks like there's not enough information for us to know how to help you.

If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

Requesting a rule change? Please see [Proposing a Rule Change](http://eslint.org/docs/developer-guide/contributing/rule-changes) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by ilyavolodin** [comment]
@dozoisch Could fill out our [Proposing a Rule Change](http://eslint.org/docs/developer-guide/contributing/rule-changes) template please? That makes it a lot easier for us to triage issues.

**Comment by dozoisch** [comment]
**What rule do you want to change?**
no-restricted-imports, no restricted-modules

**Does this change cause the rule to produce more or fewer warnings?**
No, different warnings

**How will the change be implemented? (New option, new default behavior, etc.)?**
Following the same format as http://eslint.org/docs/rules/no-restricted-properties. The config instead of being an array of string will be an array of string OR object containing "name", "message".* This way you can specify a custom message for module.

**Please provide some example code that this change will affec**t:

```js
// "no-restricted-modules": [2, { name: "react-bootstrap", message: "...."}]
const RB = require('react-bootstrap'); // warns
const Button = require('react-bootstrap/lib/button'); // passes
```


**What does the rule currently do for this code?**
Warns with a very obscure message. `'react-bootstrap' import is restricted from being used`

**What will the rule do after it's changed?**
Warn with a custom message.

---

I'd like to discuss a bit the format!

Here are a two possibilities:

```js
// Single name, single message
{
  name: 'single-name',
  message: 'message for this particular one'
}

// Multiple names per message
{ 
   names: [ 'react-bootstrap', 'lodash' ], 
   message: 'message for this category',
}
```

The first option is very verbose and might lead to duplication (of messages) but is exactly the same as no-restricted-properties

The second allows for "categories" of modules instead of duplicating the message everytime, but is a bit different from no-restricted-properties.

I'm not particularly attached to one of the two! So I'm open to suggestion. Also, I use the word name right now, but it can be changed, I'd prefer if we had the same word for both modules and imports to allow the configuration to be the same instead of having to duplicate it & change it.

**Comment by dozoisch** [comment]
@ilyavolodin Let me know what you think!

**Comment by ljharb** [comment]
My preference is "single name, single message", for simplicity, but either seem fine.

**Comment by ilyavolodin** [comment]
@dozoisch Thanks for doing that. I would be OK with modifying this rule to follow schema used in `no-restricted-properties` rule (more or less).

**Comment by dozoisch** [comment]
Perfect @ilyavolodin! I can follow the same schema. How would you name the "name" property earlier? I went with name because "module" doesn't really apply for -import rule and vice-versa.

**Comment by ilyavolodin** [comment]
I think `name` is fine. I would skip multiple `names` though. Just single string is fine for most cases, I think.

**Comment by dozoisch** [comment]
@ilyavolodin Alright! Yes, it'll support both (old and new)! I should have some time to do it today or tomorrow.

**Comment by dozoisch** [comment]
@ljharb @ilyavolodin So I started looking into it. For pattern support the code uses the package "ignores". The problem is that ignores doesn't tell you which rule worked or didn't work since you can provide negative rules that can override each others. I'm not sure what to do for these. 

For the rules that have direct match, it's pretty easy, I can use the set just like now.

For the patterns... whether I change what is use to test it (as ignores only supports returning a boolean true or false). But that could be a breaking change. Or I can have the custom message work only for normal paths, not for patterns.

What are your thoughts?

**Comment by nzakas** [comment]
Team - we need a champion for this otherwise it's time to close the issue.

**Comment by kaicataldo** [comment]
I can champion this. I think this is a really useful enhancement and it puts no-restricted-imports more in line with no-restricted-properties.

**Comment by kaicataldo** [comment]
@eslint/eslint-team Anyone else willing to support this? We need two more thumbs.

**Comment by mikesherov** [comment]
:+1:

**Comment by not-an-aardvark** [comment]
:+1:

**Comment by kaicataldo** [comment]
@dozoisch Apologies for the delay - are you still able and willing to work on this?

**Comment by dozoisch** [comment]
@kaicataldo Hey! Yes, I'm still willing to work on this, if you can guide me in my previous interrogations! https://github.com/eslint/eslint/issues/7532#issuecomment-260100106

**Comment by kaicataldo** [comment]
@dozoisch Hey, sorry, I'm having difficulty understanding the question. Mind trying to explain it again, maybe with some examples?

**Comment by dozoisch** [comment]
@kaicataldo Will do tonight

**Comment by dozoisch** [comment]
@kaicataldo 

So the rule currently supports two way of specifying paths.

You can specify "paths" which is a direct match. (ie: does indexOf()).

You can specify patterns. Now the issues is that patterns can be additive or negative. They aren't direct matches. They use `ignore` which follows the gitignore spec) which means we can't really pinpoint which rule would be matching.

So the new schema would be something like 

```js
{
  name: 'single-name', // TBD for the param (could be path also)
  message: 'message for this particular one'
}
```

The problem comes with pattern support.

I guess we could allow user passing an array in pattern and we test them separately (each group), so this way we can pinpoint  a specific group.

```js
{
  pattern: [...],
  message: 'message for this pack'
}
```

---


So full schema would be 

```js
type OldSchema =  { paths?: Array<string>, patterns?: Array<string> };
type NewSchema = { path?: string, pattern?: Array<string>, message?: string };
type Severity = string | number; // 'error', 'warning', etc
type Args = [ OldSchema ] | Array<NewSchema>; 
[ Severity, ...Args ];
```

New schema supports path or pattern. The singular name could help us know if it's new or not so we could easily throw error for the config if needed.

When message is null we provide a default one.

What do you think of that?

**Comment by dozoisch** [comment]
@kaicataldo Let me know, I could start with that this week probably.

**Comment by kaicataldo** [comment]
@dozoisch Maybe I'm just missing something, but I'm not understanding what the issue is. I think we should stick to one path/pattern per specific object (so no `names` or `patterns`, as mentioned above).

There's been a lot of discussion since the original proposal - would you mind updating it to show what the state of the proposal is now? Sorry, having trouble following.

**Comment by dozoisch** [comment]
@kaicataldo Problem is sticking to only 1 pattern per block will make it a lot less powerful. Patterns are not independent, and can be used as an array (from the ignore module) to have very specific ignores (as you can have "negatives" ignores). It should stay as an array for each thing. Else we will potentially not support some use cases that are currently supported. 

There is no issue anymore, I found a way, and my post was the solution itself. And I think that solution would work.

**Comment by dozoisch** [comment]
@kaicataldo Alright, sorry for the delay! I updated https://github.com/eslint/eslint/issues/7532#issuecomment-258496035

**Comment by Aghassi** [comment]
@kaicataldo @dozoisch Any movement on this? I was looking for ways to support this sort of functionality for some tools I'm writing and would be interested in helping if this still needs doing. Alternatively, if there is a better way to do this sort of thing, I'm all ears. Especially at scale with Webpack, tree shaking is quite useful if enforced.

**Comment by nzakas** [comment]
@Aghassi Updates are always posted on the issue, so no need to ask if there are updates. :)

As it stands, this issue is considered stale, and as such, will be closed unless someone volunteers to implement the change.

**Comment by nzakas** [comment]
Unfortunately, it looks like there wasn't enough interest from the team or community to implement this change. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to be implemented after 90 days tend to never be implemented, and as such, we [close those issues](https://eslint.org/docs/maintainer-guide/issues#when-to-close-an-issue). This doesn't mean the idea isn't interesting or useful, just that it's not something the team can commit to.
